server/protocols/jmap/src/main/java/org/apache/james/jmap/AuthenticationFilter.java

```diff
@@ -66,43 +69,34 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         HttpServletRequest httpRequest = (HttpServletRequest) request;
         HttpServletResponse httpResponse = (HttpServletResponse) response;
 
-        Optional<String> authHeader = Optional.ofNullable(httpRequest.getHeader("Authorization"));
-        if (!checkAuthorizationHeader(authHeader)) {
+        try {
+            HttpServletRequest requestWithSession = authMethods.stream()
+                    .filter(auth -> auth.checkAuthorizationHeader(getAuthHeaders(httpRequest)))
+                    .findFirst()
+                    .map(auth -> addSessionToRequest(httpRequest, createSession(auth, getAuthHeaders(httpRequest))))
+                    .orElseThrow(UnauthorizedException::new);
+            chain.doFilter(requestWithSession, response);
+
+        } catch (UnauthorizedException | NoAuthHeaderException | MailboxCreationException | JwtException e) {
+            LOGGER.error("Exception occurred during authentication process", e);
             httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED);
-            return;
         }
 
-        addSessionToRequest(httpRequest, httpResponse, authHeader);
-
-        chain.doFilter(httpRequest, response);
     }
 
-    private void addSessionToRequest(HttpServletRequest httpRequest, HttpServletResponse httpResponse, Optional<String> authHeader) throws IOException {
-        try {
-            MailboxSession mailboxSession = createMailboxSession(authHeader);
-            httpRequest.setAttribute(MAILBOX_SESSION, mailboxSession);
-        } catch (MailboxException e) {
-            httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED);
-        }
+    private Stream<String> getAuthHeaders(HttpServletRequest httpRequest) {
+        Enumeration<String> authHeaders = httpRequest.getHeaders(AUTHORIZATION_HEADERS);
+
+        return authHeaders != null && authHeaders.hasMoreElements() ? Collections.list(authHeaders).stream() : Stream.of();
```

Comment:
OK, fixed. Can't find the test fixture that used mocked request and failed because of the missing assertion (hasMoreElements).